### ISSUE TITLE
Random 404s on homepage after login

### ISSUE THREAD

**Issue by dmandras** [Apexal/late#392]
When logging in, there is a small chance that the main page will have a 404 for the content, along with the weblink `https://www.late.work/auth/unidentified` in the browser. 

This has so far only happened to me on iOS 13 Dev Beta 8, iPhone XS and Safari. 

Screencapture of the whole page:
![Image](https://i.imgur.com/zWZ8ieq.jpg)

**Comment by Apexal** [comment]
I made some changes that *might* fix this. Let me know if it happens again!

**Comment by Apexal** [comment]
I've now had a handful of people point this out. Something is up...

**Comment by robere2** [comment]
@Apexal Were you able to make any progress on this? I managed to capture the requests that caused this in a HAR file while trying to fix another bug, which might give more insight. They do have my RPI credentials so I can't distribute it here obviously. I haven't dug through it yet to see if anything is out of the ordinary.

**Comment by robere2** [comment]
Seems to occur when `https://www.late.work/auth/login` responds with 302 to location `undefined`.
Most likely is stemming from a bug in the [koa2-cas](https://www.npmjs.com/package/koa2-cas) module, or some misuse on LATE's end. Reproducable in browsers that haven't visited Late before (Incognito works too), however interestingly this doesn't apply to Chrome.

**Comment by Apexal** [comment]
`sakaib` reached out to me on Mattermost and revealed that this bug occurred and gave him access to the site when he was still on the waitlist! This explains how students on the waitlist have been able to temporarily access the page.

This is bad. We need to fix this.

**Comment by robere2** [comment]
> `sakaib` reached out to me on Mattermost and revealed that this bug occurred and gave him access to the site when he was still on the waitlist! This explains how students on the waitlist have been able to temporarily access the page.
> 
> This is bad. We need to fix this.

Can you post his message here?

**Comment by Apexal** [comment]
the full message I was sent:
> Hey. I understand you're the team lead for LATE. I was playing with cas trying to understand how it works and I think I did something that took me off the wait list. Thought you might be interested in taking a look. RCSID is sakaib and it was a few minutes ago if you're looking at server logs. I think it had something to do with landing on /auth/undefined after receiving my cas ticket and that being handled incorrectly. Lmk if you want more info.

**Comment by robere2** [comment]
Lately this has been happening to me in combination of #124 every single time I log in. I'm using the Edge beta browser. In order to get out of the redirect loop I click on the URL bar and press enter in order to load the CAS page from scratch again.

**Comment by robere2** [comment]
I did some digging. I believe it to be a bug/oversight in [koa2-cas](https://github.com/Donny2333/koa2-sso), or rather a conflict between it and and our session manager, or how we implemented it.
koa2-cas follows these steps:
1. call `_login()` which sets `ctx.session.cas_return_to` to `ctx.query.returnTo` or `url.parse(ctx.url).path`, then redirects the user to the CAS login page
2. Wait for the user to return to the page, then call `_handleTicket()` if they have a ticket in the query string (which they will).
3. Verify the CAS ticket.
4. If CAS ticket is invalid, send 401 response. If valid, send 302 response, pointing to `ctx.session.cas_return_to`.

This is great and all, however if you inspect the requests, the session tokens are not the same throughout the entire process. I suspect koa-session or we are changing them for some reason. 

On the **first** request to https://late.work/auth/login, a `koa:sess` cookie is sent by the browser. The server responds with a 302 response, `Location` header pointing to CAS, and a `Cookie` header is not included in the response, so nothing is changed in the browser.

On the **second** request to https://late.work/auth/login (the one that includes the ticket query parameter), the same `koa:sess` cookie is sent from the browser to the server. A 302 response is sent with a `Location` header pointing to `undefined`. But this time a `Cookie` header IS included in the response, changing `koa:sess` to a different value. 

My suspicion is before koa2-cas can view the session data, the session is deleted and a new session is created for some reason. Therefore the `cas_return_to` session value is undefined. I haven't been able to test or identify the cause for #124 yet however I'm going to try to figure it out.
